[29기 박겸영] Add: 메인 페이지 레이아웃

seoltang opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [x] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 메인 페이지 레이아웃 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. Home 컴포넌트
- 메인 페이지에 가장 먼저 보이는 첫 번째 section을 Home 컴포넌트로 만들었습니다.
2. TwoColumnsPromotion 컴포넌트
- TwoColumnsPromotion 컴포넌트와 LinkButton 컴포넌트는 다른 페이지에서도 사용되기 때문에 공용 components 폴더에 포함했습니다.
- 텍스트가 왼쪽/오른쪽에 있을 때 레이아웃을 다르게 설정하기 위해 `isTextOnLeft`를 props로 주고 불리언 값에 따라 왼쪽/오른쪽 정렬이 되도록 구현했습니다.
- 컴포넌트 내 aside에 대부분 이미지가 들어가지만, 다른 태그가 들어가는 경우도 있어서 모든 경우에 재사용하기 위해 `isContentImg` props를 주고 true이면 `<img>` 태그가 들어가도록, false이면 `props.children`이 들어가도록 설계했습니다.

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- 이미지 등 문서의 주요 흐름과는 독립적인 콘텐츠를 표현하는 `<figure>` 태그와 반응형 이미지를 구현하기 위해 필요한 `<picture>` 태그를 알게 되었습니다. 이후에 반응형 레이아웃을 구현할 때 더 유용하게 활용될 것 같습니다.
- Home 컴포넌트의 배경에 보이는 이미지를 반응형으로 구현하는 것이 까다로웠습니다. 화면 크기가 변경되어도 이미지가 이상하게 보이지 않도록 ①이미지와 텍스트 전체를 감싸는 부모 태그에 `display: flex` 속성을 설정하고, ②이미지 태그에 `height: 100%; object-fit: cover;` 속성을 설정했습니다.
- TwoColumnSection 컴포넌트를 재사용할 때 특정 컴포넌트에만 태그를 추가하기 위해서 props.children을 사용해 보았습니다.

<br />

## :: 기타 질문 및 특이 사항
- props.children을 컴포넌트 안의 여러 부분에 넣고 싶을 때는 어떻게 하는지 궁금합니다.

seoltang committed:
Add: Main page: first section

seoltang committed:
Add: 메인 페이지 레이아웃
- LinkButton 컴포넌트 공용 폴더에 추가
- TwoColumnSection 컴포넌트 공용 폴더에 추가
- 메인 페이지 레이아웃(TwoColumnSection) 추가

seoltang committed:
Modify: variables.scss: 공용 변수 추가

201steve commented on src/pages/Main/Main.js:
```diff
@@ -1,7 +1,40 @@
+  return (
+    <>
+      <Home />
+      <TwoColumnSection
+        isTextOnLeft={true}
+        h2text="스킨 케어 루틴을 관리하는 방법"
+        content="세심하고 사려 깊은 스킨 케어 루틴에는 우리 피부만큼이나 소중한 '시간'이 필요하며 스스로를 보살피는데 사용하는 시간은 언제나 유익합니다."
+        btnText="가이드 읽기"
+        btnLink="/"
+        src="/images/main/main-2.jpg"
+      >
+        <h3 className="subTitle">실험실에서 배운 교훈</h3>
+      </TwoColumnSection>
```
* 반복된 UI는 데이터 구조 설계하고 .map 함수로 구현 해 보세요! 
* children 사용 👍🏻

201steve commented on src/pages/Main/Main.js:
```diff
@@ -1,7 +1,40 @@
+      >
+        <h3 className="subTitle">실험실에서 배운 교훈</h3>
+      </TwoColumnSection>
+      <TwoColumnSection
+        isTextOnLeft={false}
+        h2text="놀라움과 즐거움을 전하는 기프트"
+        content="관대함을 표현하고 싶거나 축하할 일이 생겼을 때, 영감이 충만할 때, 위솝의 기프트 가이드는 받으시는 모든 분들에게 적합한 다양한 제품을 소개해 드립니다."
+        btnText="모든 기프트 보기"
+        btnLink="/"
+        src="/images/main/main-3.jpg"
+      />
```
변수 이름 지을때 제 3자가 봐도 예상할 수 있고 직관적인 이름을 지어주시는게 좋습니다. 
* 예를 들면, h2Text는 제목 텍스트 일 것 같은데 어떤 제목 텍스트 인지 예상이 잘 안되고,
* content도 어떤 content 인지 추상적입니다.
* 컴포넌트이름인 TwoColumnSection도 컬럼이 2개인 섹션 으로만 보여서, 어떤 형태일지는 예상이되지만 어떤UI 일지는 예상이 조금 어렵네요.
* 직관적이면서 구체적으로 예상 가능한 네이밍이 어렵지만, 이 기회에 연습해보면 좋겠어요 😄

201steve commented on src/pages/Main/Home/Home.js:
```diff
@@ -0,0 +1,38 @@
+  return (
+    <section className="Home">
+      <figure className="homeImg">
+        <picture>
+          <img alt="Healthy" src="/images/main/main-1-wide.jpg" />
+        </picture>
```
이부분은 참고만 해주세요!
* picture 태그는 img 태그와 더불어 source 태그의 활용도 공부해보면 좋겠습니다!
* [반응형 이미지 참고자료](https://developer.mozilla.org/ko/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images)

201steve commented on src/components/TwoColumnSection/TwoColumnSection.js:
```diff
@@ -0,0 +1,25 @@
+import LinkButton from '../LinkButton/LinkButton';
+import './TwoColumnSection.scss';
+
+const TwoColumnSection = props => {
+  const { isTextOnLeft, children, h2text, content, btnText, btnLink, src } =
+    props;
```
* 매개변수 자리에서 바로 객체구조분해 할당 가능합니다

seoltang replied:
> * 매개변수 자리에서 바로 객체구조분해 할당 가능합니다

원래 매개변수 자리에서 했었는데 props가 많아지니까 줄바꿈이 되어 보기 불편한 것 같아서 아래에 따로 변수선언을 했습니다. props가 많은 경우에도 매개변수 자리에서 구조분해할당을 하는 것이 좋은가요?

201steve replied:
* 네!겸영님 props가 많아지면 prettier 설정에따라 멀티라인으로 줄바꿈되는건 똑같습니다.
* 컴포넌트가 전달해주는 props가 많아질때 한줄로 길게 늘어뜨려진걸 보는거랑, 줄바꿈되서 어떤 props가 있는지 한눈에 보는거랑 비교해보면 어느쪽이 더 가독성 좋은 코드인지 느끼실 수 있습니다.

201steve reviewed (commented):
첫 PR 고생많으셨습니다. 
리뷰 내용 확인하시고 적용 해 주세요 😄

seoltang committed:
Merge branch 'feature/common' into feature/main

seoltang committed:
Add: 메인 페이지 캐러셀 구현 중

seoltang committed:
Refactor: PR 리뷰 반영
- 변수명, 컴포넌트명 수정
- `<TwoColumnsPromotion>` 컴포넌트 구조 변경
  - aside 콘텐트가 이미지인지, 이미지가 아닌지에 따라 props 변경
- `<TwoColumnsPromotion>` 관련 상수데이터 생성 및 `map()` 함수 사용

201steve commented on src/pages/Main/Carousel/Carousel.js:
```diff
@@ -0,0 +1,72 @@
+  );
+};
+
+const CAROUSEL_PRODUCTS = [
+  {
+    idx: 0,
+    heading: '파슬리 씨드 안티 옥시던트 페이셜 트리트먼트',
+    description: '집중적인 보습으로 피부를 탄탄하게 강화시킵니다.',
+    src: 'https://images.unsplash.com/photo-1628088306391-1631cdd36e7c?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=580&q=80',
+  },
+  {
+    idx: 1,
+    heading: '파슬리 씨드 안티 옥시던트 페이셜 트리트먼트',
+    description: '집중적인 보습으로 피부를 탄탄하게 강화시킵니다.',
+    src: 'https://images.unsplash.com/photo-1628088306391-1631cdd36e7c?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=580&q=80',
+  },
+  {
+    idx: 2,
+    heading: '파슬리 씨드 안티 옥시던트 페이셜 트리트먼트',
+    description: '집중적인 보습으로 피부를 탄탄하게 강화시킵니다.',
+    src: 'https://images.unsplash.com/photo-1628088306391-1631cdd36e7c?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=580&q=80',
+  },
+  {
+    idx: 3,
+    heading: '파슬리 씨드 안티 옥시던트 페이셜 트리트먼트',
+    description: '집중적인 보습으로 피부를 탄탄하게 강화시킵니다.',
+    src: 'https://images.unsplash.com/photo-1628088306391-1631cdd36e7c?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=580&q=80',
+  },
+  {
+    idx: 4,
+    heading: '파슬리 씨드 안티 옥시던트 페이셜 트리트먼트',
+    description: '집중적인 보습으로 피부를 탄탄하게 강화시킵니다.',
+    src: 'https://images.unsplash.com/photo-1628088306391-1631cdd36e7c?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=580&q=80',
+  },
+];
```
* 만약 다른 컴포넌트에서 이 데이터에 접근한다면, 이 컴포넌트 파일에 접근해서 상수데이터를 꺼내와야 합니다.
* 가독성과 유지보수를 쉽고 편하게 하기위해서는 이런 데이터는 파일로 분리해서 관리하는게 좋습니다.
* 파일로 분리 해 주세요

seoltang committed:
Refactor: props 구조분해할당 parameter 위치로 변경

201steve commented on src/pages/Main/Carousel/Carousel.js:
```diff
@@ -0,0 +1,72 @@
+import React from 'react';
+import './Carousel.scss';
+
+const CarouselProduct = ({ heading, description, src }) => {
+  return (
+    <div className="CarouselProduct">
+      <div className="productImg">
+        <img alt="product" src={src} />
+      </div>
+      <div className="textWrapper">
+        <h5>{heading}</h5>
+        <div className="description">{description}</div>
+      </div>
+    </div>
+  );
+};
```
* 이 컴포넌트도 파일로 분리 해 주세요.
* 컴포넌트는 재사용 가능하기 때문에, 가능한 쪼개고, 분리하고, 따로 관리해야 
* 유지보수하기 좋고, 코드 구조가 어떻게 이루어져있는지 한눈에 파악하기 쉽기 때문이에요.

201steve commented on src/pages/Main/Main.js:
```diff
@@ -1,7 +1,71 @@
+  );
 };
 
+const PROMOTIONS = [
+  {
+    idx: 0,
+    isTextOnLeft: true,
+    isContentImg: true,
+    subTitle: '실험실에서 배운 교훈',
+    heading: '스킨 케어 루틴을 관리하는 방법',
+    paragraph:
+      "세심하고 사려 깊은 스킨 케어 루틴에는 우리 피부만큼이나 소중한 '시간'이 필요하며 스스로를 보살피는데 사용하는 시간은 언제나 유익합니다.",
+    btnText: '가이드 읽기',
+    btnLink: '/',
+    alt: 'skincare',
+    src: '/images/main/main-2.jpg',
+  },
+  {
+    idx: 1,
+    isTextOnLeft: false,
+    isContentImg: true,
+    heading: '놀라움과 즐거움을 전하는 기프트',
+    paragraph:
+      '관대함을 표현하고 싶거나 축하할 일이 생겼을 때, 영감이 충만할 때, 위솝의 기프트 가이드는 받으시는 모든 분들에게 적합한 다양한 제품을 소개해 드립니다.',
+    btnText: '모든 기프트 보기',
+    btnLink: '/',
+    alt: 'gifts',
+    src: '/images/main/main-3.jpg',
+  },
+];
```
* 이부분도 파일로 분리할 수 있겠네요!

201steve reviewed (commented):
수고하셨습니다! 
리뷰 적용 해 주세요!

seoltang committed:
Refactor: PR 리뷰 반영
- 상수데이터 mainData.js 파일로 분리
- CarouselProduct 컴포넌트 파일로 분리
